File: 卡码/代码随想录/Python/二叉树/05-二叉树公共祖先/LeetCode_0236.py
#236. 二叉树的最近公共祖先
from typing import *
from ..TreeNode import *

class Solution:
  def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
    if not root or root==p or root==q:
      return root
    left:TreeNode=self.lowestCommonAncestor(root.left,p,q)
    right:TreeNode=self.lowestCommonAncestor(root.right,p,q)
    if left and right:
      return root
    elif not left and right:
      return right
    elif left and not right:
      return left
    else:
      return None

File: 卡码/代码随想录/C++/二叉树/05-二叉树公共祖先/LeetCode_0236.cpp
//236. 二叉树的最近公共祖先
#include <iostream>
#include "../TreeNode.h"
using namespace std;

class Solution {
public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if(!root||root==p||root==q) return root;
    TreeNode* left=lowestCommonAncestor(root->left,p,q);
    TreeNode* right=lowestCommonAncestor(root->right,p,q);
    if(left&&right) return root;
    else if(!left&&right) return right;
    else if(left&&!right) return left;
    else return nullptr;
  }
};

int main(){
  system("pause");
  return 0;
}
